Functional.AsyncEnumerables/AsyncEnumerator.cs

```diff
@@ -1,31 +1,35 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Text;
+﻿using System.Collections.Generic;
 using System.Threading.Tasks;
 
 namespace Functional
 {
-	internal class AsyncEnumerator<T> : IAsyncEnumerator<T>
-    {
+	internal class AsyncEnumerator<T> : DisposableBase, IAsyncEnumerator<T>
+	{
 		private static readonly Task<bool> _trueResult = Task.FromResult(true);
 
 		private static readonly Task<bool> _falseResult = Task.FromResult(false);
 
-		private readonly Lazy<Task<IEnumerator<T>>> _enumerator;
+		private readonly AsyncLazy<IEnumerator<T>> _enumerator;
 
-		public T Current => _enumerator.IsValueCreated && _enumerator.Value.IsCompleted ? _enumerator.Value.Result.Current : default;
+		public T Current => GetCurrentAsync().ConfigureAwait(false).GetAwaiter().GetResult();
 
-		internal AsyncEnumerator(Lazy<Task<IEnumerator<T>>> enumerator)
+		internal AsyncEnumerator(AsyncLazy<IEnumerator<T>> enumerator)
 			=> _enumerator = enumerator;
 
-		public Task<bool> MoveNext()
+		public async Task<bool> MoveNext()
 		{
-			var enumerator = _enumerator.Value;
+			var enumerator = await _enumerator.GetValueAsync().ConfigureAwait(false);
```

Comment:
I believe this will be more expensive to evaluate than the old code, since in the old code, all calls after the first early out synchronously. The new code here involves many additional function calls and continuations (although admittedly they will typically be synchronous).

Reply:
@JohannesMoersch In the AsyncLazy implementation, only first call to this async method is asynchronous, which is guaranteed by an atom InterLocked operation, all subsequent calls are synchronous with no overhead at all.  Besides that,  async await pattern makes the thread reusable when appropriate while Task.Wait or Task.GetResult will always block the thread.

Reply:
There are still significantly more function calls per iteration in the new code. The old code never blocked as well. Was there any particular issue you were addressing with the old code?

Reply:
VSThreading analyzer warns  Lazy<Task<T>> may causes deadlock

Reply:
You can apply the analyzers locally and see all the original warnings I try to fix here

Reply:
I'm taking a look, and it's incorrect about the call to .Result since I first check if the task is complete. Also, it's concern about using the current TaskScheduler was actually intentional on my part. You probably noticed that I do not use .ConfigureAwait(false) anywhere throughout my code. I did this because otherwise the behaviour of delegates passed into the system is undefined. If the library is used in WinForms or WPF (with a defined UI thread), delegates provided by the UI thread will sometimes execute on the UI thread (if the continuation occurs synchronously) and sometimes execute on the thread pool. This can introduce race conditions. In particular, if the delegate accesses properties on a UI element, they will throw exceptions when they execute on the thread pool. The end result is that the library is unsafe to use with .Wait() or .Result calls on tasks, but software shouldn't be developed this way anyways. I prefer this to the undefined behaviour within the delegates themselves.

Reply:
Ah, you are right. The original implementation does not wait on the Task, however, it requires the upstream caller to explicitly wait on the task before calling the Current getter, otherwise, it gets default T value, which is not intuitive to me.  In my opinion, it should either automatically perform the wait, or throws clear exceptions instead of giving default T to the caller

Reply:
I think the original design to get rid of task wait indeed makes this class itself clean and safe, but it just delegates the problem to the caller, and there has to be someone doing the dirty work somewhere else, which does not seem to be ideal to me either

Reply:
Whomever calls MoveNext needs to await it anyways, so the behaviour is actually the same. async functions that await calls are actually just syntactic sugar over multiple functions wired together with continuations similar to the code in the original method. I'd prefer that Current not return any data in MoveNext has not been awaited, since that is an incorrect usage of the class.